In [3]:
# Import required libraries and dependencies
import os
os.environ['OMP_NUM_THREADS'] = '1'
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler 



In [4]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [5]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [7]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [8]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data)

In [9]:
# crypto_db = pd.DataFrame({
#     "Price Change % (24h)": crypto_scaled['price_change_percentage_24h']
# })

In [10]:
# Create a DataFrame with the scaled data
crypto_db = pd.DataFrame(crypto_scaled, columns=['price_change_percentage_24h', 
                                                 'price_change_percentage_7d', 
                                                 'price_change_percentage_14d', 
                                                 'price_change_percentage_30d', 
                                                 'price_change_percentage_60d', 
                                                 'price_change_percentage_200d',	
                                                 'price_change_percentage_1y'])

# Copy the crypto names from the original data
coin_names = list(df_market_data.index)

# Set the coinid column as index
crypto_db['coin_id'] = coin_names  
crypto_db = crypto_db.set_index('coin_id') 

# Display sample data
crypto_db.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [11]:
# Create a list with the number of k-values from 1 to 11
k_list = list(range(1,12))
k_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [12]:
# Create an empty list to store the inertia values
inertia_list = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_list:

    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=k, random_state=1, n_init=10)

    # 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(crypto_db)

    # 3. Append the model.inertia_ to the inertia list
    inertia_list.append(k_model.inertia_)

In [13]:
# Create a dictionary with the data to plot the Elbow curve
graph_data = {"k": k_list, "inertia": inertia_list}

# Create a DataFrame with the data to plot the Elbow curve
graph_df = pd.DataFrame(graph_data)
graph_df.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   63.858668

In [14]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
graph_df.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow Curve for k-Value vs Inertia",
    xticks=k_list,
    color="#3b735e"
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for k is 4.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [15]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1, n_init=10)

In [16]:
# Fit the K-Means model using the scaled data
model.fit(crypto_db)

KMeans(n_clusters=4, n_init=10, random_state=1)

In [17]:
# Predict the clusters to group the cryptocurrencies using the scaled data
crypto_cluster = model.predict(crypto_db)

# Print the resulting array of cluster values.
crypto_cluster

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [18]:
# Create a copy of the DataFrame
db_copy = crypto_db.copy()

In [19]:
# Add a new column to the DataFrame with the predicted clusters
db_copy['predicted_cluster']=crypto_cluster

# Display sample data
db_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_cluster  
coin_id                                                      
bitcoin                        -0.251637                  2  
ethereum                       -0.199352                  2  
tether                         -0.282061                  0  
ripple                         -0.295546                  0  
bitcoin-cash                   -0.270317                  2

In [20]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
db_copy.hvplot.scatter(
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
    by='predicted_cluster',
    hover_cols = 'coin_id'
)

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [24]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [25]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_cluster = pca.fit_transform(db_copy)

# View the first five rows of the DataFrame. 
pca_cluster[:5]

array([[ 0.44890795, -1.24537573, -0.8506404 ],
       [ 0.49536709, -0.899823  , -1.31755943],
       [-0.81884571,  0.07189909,  0.69501507],
       [-0.84035711,  0.08005407,  0.54435952],
       [ 0.81324004, -2.66952035, -1.64332113]])

In [27]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.34871677, 0.31363391, 0.22627118])

In [28]:
#Finding the sum of explained variances
sum(pca.explained_variance_ratio_)

0.8886218549859446

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.34871677 + 0.31363391 + 0.22627118 = 0.8886218549859446

In [30]:
# Create a new DataFrame with the PCA data.
pca_cluster_df = pd.DataFrame(pca_cluster, columns = ["PCA1", "PCA2", "PCA3"])

# Creating a DataFrame with the PCA data
# Copy the crypto names from the original data
pca_cluster_df['coin_id'] = list(db_copy.index)

# Set the coinid column as index
pca_cluster_df = pca_cluster_df.set_index('coin_id')

# Display sample data
pca_cluster_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin       0.448908 -1.245376 -0.850640
ethereum      0.495367 -0.899823 -1.317559
tether       -0.818846  0.071899  0.695015
ripple       -0.840357  0.080054  0.544360
bitcoin-cash  0.813240 -2.669520 -1.643321

---

### Find the Best Value for k Using the PCA Data

In [31]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [32]:
# Create an empty list to store the inertia values
inertia_list2 = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:

    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1, n_init=10)

    # 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(pca_cluster_df)

    # 3. Append the model.inertia_ to the inertia list
    inertia_list2.append(k_model.inertia_)

In [33]:
# Create a dictionary with the data to plot the Elbow curve
graph_data2 = {"k": k, "inertia": inertia_list2}

# Create a DataFrame with the data to plot the Elbow curve
graph_df2 = pd.DataFrame(graph_data2)
graph_df2

k     inertia
0    1  290.622694
1    2  203.579726
2    3  112.854846
3    4   44.130602
4    5   33.217427
5    6   23.371677
6    7   17.172200
7    8   13.593411
8    9   11.155098
9   10    8.958913
10  11    7.345260

In [42]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
graph_df2.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow Curve for k-Value vs Inertia - PCA Value",
    xticks=k,
    color="#1e54d4"
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k using PCA data was `4`
___
* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, the best k-value was still the same

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [36]:
# Initialize the K-Means model using the best value for k
model_2 = KMeans(n_clusters=4, random_state=1, n_init=10)

In [37]:
# Fit the K-Means model using the PCA data
model_2.fit(pca_cluster_df)

KMeans(n_clusters=4, n_init=10, random_state=1)

In [39]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4 = model_2.predict(pca_cluster_df)

# Print the resulting array of cluster values.
k_4

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 3, 1, 0, 0, 2, 0, 0, 0, 0])

In [40]:
# Create a copy of the DataFrame with the PCA data
pca_df_copy = pca_cluster_df.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_df_copy['predicted_cluster'] = k_4

# Display sample data
pca_df_copy.head()

PCA1      PCA2      PCA3  predicted_cluster
coin_id                                                      
bitcoin       0.448908 -1.245376 -0.850640                  1
ethereum      0.495367 -0.899823 -1.317559                  1
tether       -0.818846  0.071899  0.695015                  0
ripple       -0.840357  0.080054  0.544360                  0
bitcoin-cash  0.813240 -2.669520 -1.643321                  1

In [41]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_df_copy.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by = 'predicted_cluster',
    hover_cols='coin_id'
)

:NdOverlay   [predicted_cluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [45]:
# Composite plot to contrast the Elbow curves
graph_df.hvplot.line(x="k", y="inertia", title= "Elbow Curve - Original Data", xticks=k_list, color="#3b735e") + \
    graph_df2.hvplot.line(x="k", y="inertia", title="Elbow Curve - PCA Values", xticks=k, color="#1e54d4")

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [46]:
# Composite plot to contrast the clusters
db_copy.hvplot.scatter(x='price_change_percentage_24h', y='price_change_percentage_7d', by='predicted_cluster', hover_cols = 'coin_id') + \
pca_df_copy.hvplot.scatter(x="PCA1", y="PCA2", by = 'predicted_cluster', hover_cols='coin_id')

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted_cluster]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The graph formed tighter clusters after using just the PCA Values. There are also more entries under cluster 1 than the original analysis.